CMake/Findmlpack.cmake

```
@@ -0,0 +1,641 @@
+    find_cereal(${CMAKE_BINARY_DIR})
+    if (CEREAL_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${CEREAL_INCLUDE_DIR})
+    endif()
+  endif()
+
+  mlpack(${CMAKE_BINARY_DIR})
+  if (NOT MLPACK_FOUND)
+    get_deps(https://www.mlpack.org/files/mlpack-latest.tar.gz mlpack mlpack-latest.tar.gz)
+    set(MLPACK_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_mlpack(${CMAKE_BINARY_DIR})
+    if (MLPACK_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${MLPACK_INCLUDE_DIR})
+    endif()
+  endif()
+
+  find_openmp() 
+
+endmacro()
+
+# Download and compile OpenBLAS if we are cross compiling mlpack for a specific
+# architecture. The function takes the version of OpenBLAS as variable.
+macro(fetch_mlpack_and_crosscompile)
+  # 1. Pull openblas and crossocompile it
+  search_openblas(0.3.28)
+  # 2. Call autodownload, DONE
+  fetch_mlpack(ON)
+endmacro()
+
+
+##===================================================
+##  MLPACK MAIN FUNCTIONS CALL. 
+##===================================================
+
+macro(find_mlpack)
+  # If we're using gcc, then we need to link against pthreads to use std::thread,
+  # which we do in the tests.
+  if (CMAKE_COMPILER_IS_GNUCC)
+    find_package(Threads)
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})
+  endif()
+
+  if (NOT MLPACK_DISABLE_OPENMP)
+    find_openmp()
+  endif ()
+
+  find_armadillo()
+  if (ARMADILLO_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${ARMADILLO_INCLUDE_DIRS})
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
+  else()
+    message(FATAL_ERROR "Armadillo not found, (required dependency of mlpack).")
+  endif ()
+
+  find_ensmallen()
+  if (ENSMALLEN_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${ENSMALLEN_INCLUDE_DIR}")
+  else()
+    message(FATAL_ERROR "Ensmallen not found, (required dependency of mlpack).")
+  endif()
+
+  find_cereal()
+  if (CEREAL_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${CEREAL_INCLUDE_DIR})
+  else()
+    message(FATAL_ERROR "Cereal not found, (required dependency of mlpack).")
+  endif()
+
+  if (USE_SYSTEM_STB)
+    find_stb()
+  endif()
+  if (StbImage_FOUND)
+    set(STB_AVAILABLE "1")
+    add_definitions(-DMLPACK_USE_SYSTEM_STB)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${STB_IMAGE_INCLUDE_DIR}")
+
+    # Make sure that we can link STB in multiple translation units.
+    include(CMake/TestStaticSTB.cmake)
+    if (NOT CMAKE_HAS_WORKING_STATIC_STB)
+      message(FATAL_ERROR "STB implementations's static mode cannot link across "
+          "multiple translation units!  Try upgrading your STB implementation, "
+          "or using the auto-downloader (set DOWNLOAD_DEPENDENCIES=ON in the "
+          "CMake configuration command.")
+    endif()
+  endif()
+
+  mlpack()
+  if (MLPACK_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${MLPACK_INCLUDE_DIR}")
+  else()
+    message(FATAL_ERROR "mlpack not found, (required dependency of itself).")
```

```suggestion
    message(FATAL_ERROR "mlpack not found!")
```
I dunno, I'm not sure the parenthetical adds much value to the error message.